metro.config.js

```diff
@@ -73,5 +73,16 @@ module.exports = {
         inlineRequires: true,
       },
     }),
+    minifierConfig: {
+      // The Metro doc is *very* unhelpful here as of 2021-12:
+      //   https://facebook.github.io/metro/docs/configuration/#minifierconfig
+      // Looking at the implementation, I *think* the "minifier config" is
+      // passed straight to the `terser` minifier, so we should use the doc
+      // for that:
+      //   https://github.com/terser/terser#minify-options
```

Comment:
FWIW you can use the [metro-minifier-terser package](https://www.npmjs.com/package/metro-minify-terser) to switch to minify v4 (which is much better than uglify-js. There is also an extemely hack version of that package to support terser v5 [here](https://www.npmjs.com/package/@shortwave/metro-minify-terser)

Reply:
It sounds like on Hermes what's going on is that you don't end up shipping anything that looks like JS code at all. In [this 2021-10 blog post](https://reactnative.dev/blog/2021/10/26/toward-hermes-being-the-default) they're explicit about that:
> Hermes’s defining feature is how it performs compilation work ahead-of-time, meaning that React Native apps with Hermes enabled ship with precompiled optimized bytecode instead of plain JavaScript source.

So any JS source is just input to that compilation step, at build time, and giving that compiler minified vs. unminified JS doesn't make much difference to the output.


> FWIW you can use the [metro-minifier-terser package](https://www.npmjs.com/package/metro-minify-terser) to switch to minify v4 (which is much better than uglify-js).

Thanks for that tip! That seems like a good thing for us to try. (I think you meant to say "terser v4" rather than "minify v4", is that right?)

Reply:
Huh, I didn't know that about Hermes. I wonder if that means our current `ExtendableError` implementation just has no way of working.

<details>
  <summary>(our `ExtendableError` implementation)</summary>

```js
/**
 * `Error`, but subclass instances have the name of the subclass at `.name`
 *
 * All of our custom error classes should extend this. For ones that don't,
 * instances will have "Error" for their `.name`, instead of "TimeoutError",
 * "NetworkError", etc. That's worse than useless because we could
 * misidentify an error while debugging, based on the understandable but
 * wrong assumption that our custom errors' names already correspond to the
 * subclasses' names out of the box.
 */
// TODO: Add linting to make sure all our custom errors extend this
export class ExtendableError extends Error {
  // Careful! Minification might make this differ from what we see in our
  // source code, which would be kind of sad. (I guess we'll find out?)
  // Don't use this for equality checks or user-facing text:
  // https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Function/name#javascript_compressors_and_minifiers.
  //
  // This will work even down a chain of subclasses, i.e.,
  // MyVerySpecificError extends MySpecificError extends ExtendableError. If
  // you call `new MyVerySpecificError(…)`, then `MyVerySpecificError`'s
  // constructor will be the value of `this.constructor` here, so the
  // instance gets the correct name.
  name: typeof Error.name = this.constructor.name;
}
```

</details>

Reply:
> (I think you meant to say "terser v4" rather than "minify v4", is that right?)

Yupp sorry about the typo (and the drive by review, but you linked to a PR I was looking at)

Reply:
>   name: typeof Error.name = this.constructor.name;

No, I expect that's fine -- the `.name` property on a function is a lot more visible than the `.toString`, and I'd expect Hermes to preserve that.

Googling for a Hermes playground, they have one: https://hermesengine.dev/playground/

Doesn't seem to have direct links to specific demo code (like the handy flow.org/try does), but if you paste the following code there and hit the "play" button:
```js
function a_function() {}
const a_variable = a_function;
print(a_variable.name);
```
you'll see that "a_function" appears in the resulting bytecode, and "a_variable" doesn't. Then if you edit the options box at the top so it reads just `-O` (instead of `-O -dump-bytecode`) and hit the "play" button, you'll see that it prints:
```
a_function
```

So that seems to confirm it -- `.name` on a function (and therefore on a class, which is a special case of a function) is preserved just fine.

Reply:
> (and the drive by review, but you linked to a PR I was looking at)

No problem at all! I appreciate the tip.

Those cross-project links that GitHub makes can feel noisy sometimes when looking at a PR. But then when they make helpful connections possible like the information you shared, that's what makes them feel worth it :smile: